src/kakarot/instructions/system_operations.cairo

```diff
@@ -509,32 +559,84 @@ namespace CallHelper {
         bitwise_ptr: BitwiseBuiltin*,
     }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
         alloc_locals;
-        // TODO: success should be taken from ctx but revert is currently just raising so
-        // TODO: writing here TRUE: with the current implementation, a reverting sub_context
-        // TODO: would break the whole computation, so if it does not, it's TRUE
-        let success = Uint256(low=1, high=0);
-        let ctx = ExecutionContext.update_sub_context(ctx.calling_context, ctx);
-        let ctx = ExecutionContext.increment_gas_used(ctx, ctx.sub_context.gas_used);
 
-        // Append contracts selfdestruct to the calling_context
-        let ctx = ExecutionContext.push_to_destroy_contracts(
-            self=ctx,
-            destroy_contracts_len=ctx.sub_context.destroy_contracts_len,
-            destroy_contracts=ctx.sub_context.destroy_contracts,
-        );
+        let is_reverted: felt = ExecutionContext.is_reverted(self=ctx);
+        let sub_context_reverted: felt = ExecutionContext.is_reverted(ctx.sub_context);
 
-        let stack = Stack.push(ctx.stack, success);
-        let ctx = ExecutionContext.update_stack(ctx, stack);
-        // ret_offset, see prepare_args
-        let memory = Memory.store_n(
-            ctx.memory,
-            ctx.sub_context.return_data_len,
-            ctx.sub_context.return_data,
-            [ctx.sub_context.return_data - 1],
-        );
-        let ctx = ExecutionContext.update_memory(ctx, memory);
+        if (sub_context_reverted != 0) {
+            SelfDestructHelper._finalize_loop(
```

Comment:
Reusing the loop in the SelfDestructHelper. Need to refactor a name to generalize usecases across concerns.